libdevice/sanitizer_utils.cpp

```diff
@@ -6,22 +6,627 @@
+    } else {
+      __spirv_ocl_printf(__asan_shadow_value, shadow_value);
+    }
+  }
+  __spirv_ocl_printf(__newline);
+}
+
+} // namespace
+
+bool __asan_region_is_value(uptr addr, int32_t as, std::size_t size,
+                            char value) {
+  if (size == 0)
+    return true;
+  while (size--) {
+    char *shadow = (char *)MemToShadow(addr, as);
+    // __spirv_ocl_printf(__check_shadow, addr, (void *)shadow);
+    if (*shadow != value) {
+      return false;
+    }
+    ++addr;
+  }
+  return true;
+}
+
+static void __asan_internal_report_save(
+    uptr ptr, int32_t as, const char __SYCL_CONSTANT__ *file, int32_t line,
+    const char __SYCL_CONSTANT__ *func, bool is_write, uint32_t access_size,
+    DeviceSanitizerMemoryType memory_type, DeviceSanitizerErrorType error_type,
+    bool is_recover = false) {
+
+  // __spirv_ocl_printf(__save_report_before);
+
+  int Expected = ASAN_REPORT_NONE;
+  int Desired = ASAN_REPORT_START;
+  if (atomicCompareAndSet(&__DeviceSanitizerReportMem.Flag, Desired,
+                          Expected) == Expected) {
+
+    // __spirv_ocl_printf(__save_report, memory_type, error_type, access_size);
+    print_shadow_memory(ptr, as);
+
+    int FileLength = 0;
+    int FuncLength = 0;
+
+    if (file)
+      for (auto *C = file; *C != '\0'; ++C, ++FileLength)
+        ;
+    if (func)
+      for (auto *C = func; *C != '\0'; ++C, ++FuncLength)
+        ;
+
+    int MaxFileIdx = sizeof(__DeviceSanitizerReportMem.File) - 1;
+    int MaxFuncIdx = sizeof(__DeviceSanitizerReportMem.Func) - 1;
+
+    if (FileLength < MaxFileIdx)
+      MaxFileIdx = FileLength;
+    if (FuncLength < MaxFuncIdx)
+      MaxFuncIdx = FuncLength;
+
+    for (int Idx = 0; Idx < MaxFileIdx; ++Idx)
+      __DeviceSanitizerReportMem.File[Idx] = file[Idx];
+    __DeviceSanitizerReportMem.File[MaxFileIdx] = '\0';
+
+    for (int Idx = 0; Idx < MaxFuncIdx; ++Idx)
+      __DeviceSanitizerReportMem.Func[Idx] = func[Idx];
+    __DeviceSanitizerReportMem.Func[MaxFuncIdx] = '\0';
+
+    __DeviceSanitizerReportMem.Line = line;
+    __DeviceSanitizerReportMem.GID0 = __spirv_GlobalInvocationId_x();
+    __DeviceSanitizerReportMem.GID1 = __spirv_GlobalInvocationId_y();
+    __DeviceSanitizerReportMem.GID2 = __spirv_GlobalInvocationId_z();
+    __DeviceSanitizerReportMem.LID0 = __spirv_LocalInvocationId_x();
+    __DeviceSanitizerReportMem.LID1 = __spirv_LocalInvocationId_y();
+    __DeviceSanitizerReportMem.LID2 = __spirv_LocalInvocationId_z();
+
+    __DeviceSanitizerReportMem.IsWrite = is_write;
+    __DeviceSanitizerReportMem.AccessSize = access_size;
+    __DeviceSanitizerReportMem.ErrorType = error_type;
+    __DeviceSanitizerReportMem.MemoryType = memory_type;
+    __DeviceSanitizerReportMem.IsRecover = is_recover;
+
+    // Show we've done copying
+    atomicStore(&__DeviceSanitizerReportMem.Flag, ASAN_REPORT_FINISH);
+  }
+}
+
+///
+/// ASAN Error Reporters
+///
+
+void __asan_report_overlapping_error(uptr dst, uptr src, size_t size,
```

Comment:
Suggest to remove these un-implemented function firstly.

Reply:
done